src/ReturnTypeProvider/AppReturnTypeProvider.php

```diff
@@ -0,0 +1,42 @@
+<?php declare(strict_types=1);
+
+namespace Psalm\LaravelPlugin\ReturnTypeProvider;
+
+use PhpParser\Node\UnionType;
+use Psalm\CodeLocation;
+use Psalm\Context;
+use Psalm\Plugin\Hook\FunctionReturnTypeProviderInterface;
+use Psalm\StatementsSource;
+use Psalm\Type\Atomic\TNamedObject;
+use Psalm\Type\Union;
+
+final class AppReturnTypeProvider implements FunctionReturnTypeProviderInterface
+{
+
+    /**
+     * @return array<string>
+     */
+    public static function getFunctionIds(): array
+    {
+        return ['app', 'resolve'];
+    }
+
+    /**
+     * @param  array<PhpParser\Node\Arg> $call_args
+     *
+     * @return ?Type\Union
+     */
+    public static function getFunctionReturnType(StatementsSource $statements_source, string $function_id, array $call_args, Context $context, CodeLocation $code_location)
+    {
+        if (!$call_args) {
+            return new Union([
+                new TNamedObject(\Illuminate\Contracts\Foundation\Application::class),
+            ]);
+        }
+
+        // @todo: this should really proxy to \Illuminate\Foundation\Application::make, but i was struggling with that
```

Comment:
I was trying to do something like 

```
return $statements_source->getCodebase()->getMethodReturnType(
            \Illuminate\Foundation\Application::class . '::make',
            $context->self,
            $call_args
        );
```

however kept running into issues around specificity.

```
| MoreSpecificReturnType      | The declared return type 'Illuminate\Routing\Redirector' for Foo::appHelperResolvesTypes is more specific than the inferred return type 'T:fn-illuminate\foundation\application::make as object' |
| LessSpecificReturnStatement | The type 'T:fn-illuminate\foundation\application::make as object' is more general than the declared return type 'Illuminate\Routing\Redirector' for Foo::appHelperResolvesTypes
```

I must be missing something here, but both should end up with the same result regardless

Reply:
> this should really proxy to \Illuminate\Foundation\Application::make

You can simulate that by having Psalm analyse a synthetic call: https://github.com/vimeo/psalm/blob/2fe8717639514834485ba354546bdc901f85aa8d/src/Psalm/Internal/Analyzer/Statements/Block/ForeachAnalyzer.php#L722-L749

Reply:
oops, this implementation  was wrong anyway. I'll make a pr fixing